src/regex_dna.rs

```diff
@@ -0,0 +1,136 @@
+// COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+// OF THE POSSIBILITY OF SUCH DAMAGE.
+
+extern crate regex;
+
+use std::io::{self, Read};
+use std::sync::Arc;
+use std::thread;
+use regex::NoExpand;
+
+macro_rules! regex { ($re:expr) => (::regex::Regex::new($re).unwrap()); }
+
+fn run(mut seq: String) -> Vec<String> {
+    let ilen = seq.len();
+
+    seq = regex!(">[^\n]*\n|\n").replace_all(&seq, NoExpand(""));
+    let seq_arc = Arc::new(seq.clone()); // copy before it moves
+    let clen = seq.len();
+
+    let seqlen = thread::scoped(move|| {
+        let substs = vec![
+            (regex!("B"), "(c|g|t)"),
+            (regex!("D"), "(a|g|t)"),
+            (regex!("H"), "(a|c|t)"),
+            (regex!("K"), "(g|t)"),
+            (regex!("M"), "(a|c)"),
+            (regex!("N"), "(a|c|g|t)"),
+            (regex!("R"), "(a|g)"),
+            (regex!("S"), "(c|g)"),
+            (regex!("V"), "(a|c|g)"),
+            (regex!("W"), "(a|t)"),
+            (regex!("Y"), "(c|t)"),
+        ];
+        let mut seq = seq;
+        for (re, replacement) in substs.into_iter() {
+            seq = re.replace_all(&seq, NoExpand(replacement));
+        }
+        seq.len()
+    });
+
+    let variants = vec![
+        regex!("agggtaaa|tttaccct"),
+        regex!("[cgt]gggtaaa|tttaccc[acg]"),
+        regex!("a[act]ggtaaa|tttacc[agt]t"),
+        regex!("ag[act]gtaaa|tttac[agt]ct"),
+        regex!("agg[act]taaa|ttta[agt]cct"),
+        regex!("aggg[acg]aaa|ttt[cgt]ccct"),
+        regex!("agggt[cgt]aa|tt[acg]accct"),
+        regex!("agggta[cgt]a|t[acg]taccct"),
+        regex!("agggtaa[cgt]|[acg]ttaccct"),
+    ];
+    let (mut variant_strs, mut counts) = (vec!(), vec!());
+    for variant in variants.into_iter() {
+        let seq_arc_copy = seq_arc.clone();
+        variant_strs.push(variant.to_string());
+        counts.push(thread::scoped(move|| {
+            variant.find_iter(&seq_arc_copy).count()
+        }));
+    }
+
+    let mut olines = Vec::new();
+    for (variant, count) in variant_strs.iter().zip(counts.into_iter()) {
+        olines.push(format!("{} {}", variant, count.join()));
+    }
+    olines.push("".to_string());
+    olines.push(format!("{}", ilen));
+    olines.push(format!("{}", clen));
+    olines.push(format!("{}", seqlen.join()));
+    olines
+}
+
+#[cfg(not(test))]
+fn main() {
+    let mut input = String::new();
+    io::stdin().read_to_string(&mut input).unwrap();
+    println!("{}", run(input).connect("\n"));
+}
+
+#[cfg(test)]
```

Comment:
Please remove tests in the bench. it is already tested by the Makefile, and it will do more code that needed for the statistics